zkevm-circuits/src/evm_circuit/util/precompile.rs

```diff
@@ -0,0 +1,91 @@
+use bus_mapping::precompile::PrecompileCalls;
+use eth_types::Field;
+use gadgets::util::Expr;
+use halo2_proofs::plonk::Expression;
+
+use crate::evm_circuit::step::ExecutionState;
+
+use super::{
+    constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+    math_gadget::BinaryNumberGadget,
+    CachedRegion,
+};
+
+#[derive(Clone, Debug)]
+pub struct PrecompileGadget<F> {
+    address: BinaryNumberGadget<F, 4>,
+}
+
+impl<F: Field> PrecompileGadget<F> {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn construct(
+        cb: &mut EVMConstraintBuilder<F>,
+        is_success: Expression<F>,
+        callee_address: Expression<F>,
+        caller_id: Expression<F>,
+        call_data_offset: Expression<F>,
+        call_data_length: Expression<F>,
+        return_data_offset: Expression<F>,
+        return_data_length: Expression<F>,
+    ) -> Self {
+        let address = BinaryNumberGadget::construct(cb, callee_address.expr());
+
+        cb.condition(address.value_equals(PrecompileCalls::Identity), |cb| {
+            cb.constrain_next_step(ExecutionState::PrecompileIdentity, None, |cb| {
```

Comment:
Are these `require_equal` necessary if we already read the call context from rw table in precompile step? I feel we only need to make sure the next step is the precompile execution step read from the precompile info by callee address.